test/e2e/replica_set_upgrade/replica_set_test.go

```diff
@@ -0,0 +1,44 @@
+package replica_set
+
+import (
+	"testing"
+	"time"
+
+	mdbv1 "github.com/mongodb/mongodb-kubernetes-operator/pkg/apis/mongodb/v1"
+	e2eutil "github.com/mongodb/mongodb-kubernetes-operator/test/e2e"
+	"github.com/mongodb/mongodb-kubernetes-operator/test/e2e/mongodbtests"
+	f "github.com/operator-framework/operator-sdk/pkg/test"
+)
+
+func TestMain(m *testing.M) {
+	f.MainEntry(m)
+}
+
+func TestReplicaSetUpgrade(t *testing.T) {
+	ctx := f.NewTestCtx(t)
+	// defer ctx.Cleanup()
+
+	// register our types with the testing framework
+	if err := e2eutil.RegisterTypesWithFramework(&mdbv1.MongoDB{}); err != nil {
+		t.Fatal(err)
+	}
+
+	mdb := e2eutil.NewTestMongoDB()
+	t.Run("Create MongoDB Resource", mongodbtests.CreateMongoDBResource(&mdb, ctx))
+	t.Run("Config Map Was Correctly Created", mongodbtests.AutomationConfigConfigMapExists(&mdb))
+	t.Run("Stateful Set Reaches Ready State", mongodbtests.StatefulSetIsReady(&mdb))
+	t.Run("MongoDB Reaches Running Phase", mongodbtests.MongoDBReachesRunningPhase(&mdb))
+	t.Run("Test Basic Connectivity", mongodbtests.BasicConnectivity(&mdb))
+	t.Run("Test Status Was Updated", mongodbtests.Status(&mdb,
+		mdbv1.MongoDBStatus{
+			MongoURI: mdb.MongoURI(),
+			Phase:    mdbv1.Running,
+		}))
+	t.Run("MongoDB is reachable", mongodbtests.IsReachableDuring(&mdb, time.Second*10,
+		func() {
+			t.Run("Test Version can be upgraded", mongodbtests.Upgrade(&mdb, "4.0.8"))
+			t.Run("Stateful Set Reaches Ready State, after Upgrading", mongodbtests.StatefulSetIsUpdated(&mdb))
+		},
+	))
```

Comment:
should we add a downgrade straight after and ensure connectivity throughout?

We could also check that at some point the MongoDB resource has been configured to`OnDelete` update strategy.